### PR TITLE
shipowners v1

### PR DESCRIPTION
**Pull request by tmtmtl30**
## About The Pull Request

adds ship owners, a system for granting people OOC control of their ship's memo, job slots, and crew. the ship owner is generally the first person to join the ship; the role is transferred if they log out, go AFK, or possess a mob with a mind different than the mind they joined the ship with (like, mindswaps should be fine, but if you ghost, or die and respawn / an admin puts you into a simplemob / you take a ghostrole or something, you won't be the shipowner anymore). functionality for changing the ship's memo, job slots, and join mode have been moved from the cryo console to the ship owner panel that the current ship owner gets.

this PR also adds an application system. the ship owner can set the ship's join mode to "apply". players who attempt to join application-only ships must have an accepted application (organized by ckey) on the ship to join it. if none is found, they are prompted to send an application including their currently-selected character name, an optional message, and optionally their ckey. the ship owner can choose to approve or deny this application.

the ship owner is additionally able to delete applications that have already been accepted or denied, transfer ship owner status to other crewmembers, and toggle their candidacy for automatically inheriting the ship owner role if the current owner goes AFK.

there are also a bunch of misc. fixes to prevent people joining the round when they shouldn't. also the maximum number of job slots for a non-officer job on a ship has been changed to the minimum of 2x the default number of slots and 3 + the default number of slots. also, adjusting the job slots only trips the 2-minute cooldown if you are adjusting above the default slot count, and has a 5 second cooldown otherwise. 

also, admins have a button to open the ship owner panel from the shuttle manipulator panel even if they're not the ship owner, in addition to newly-added buttons to open VV for the ship's docking port and overmap datum. 

borers and split personalities almost certainly fuck with the mind-detection this system uses;  #1259 was made to rectify this but i still haven't tested that PR or tested this PR with borers / split personalities because Fuck, man.

other things:
- [ ] don't merge this fucking PR until i've commented everything because this system is utterly arcane
- [ ] needs to be properly tested with borers and split personalities
- [ ] AFK detection is scuffed -- the place i put the signal is weird, and you don't get a chance to react and stop ship owner control from leaving you if you're afk for long enough (bad if you spend a lot of time without making any inputs at all)
- [ ] the ship owner can't designate a direct successor, only toggle people's eligibility
- [ ] which might be a problem because if nobody is eligible then the ship will enter a state where it's ownerless and will continue checking for an owner every few minutes until either somebody becomes eligible or somebody new joins (who will then become an owner). so you could have a closed ship that ends up permanently ownerless because the last owner turned everybody's eligibility off and logged off or went AFK. ofc you could still ahelp this
- [ ] you don't get told when making an application "this ship doesn't currently have an owner" which seems like nice info to have
- [ ] it turns out that the config settings for AFK detection are fucked because the BYOND reference says that the "inactivity" var on clients counts the number of server ticks, not the number of ds, since the client's last input. so i need to update the config files and change some of the code to reflect that
- [ ] uhhhh 
- [ ] move "determining the next ship owner" out of O(n^2)

## Why It's Good For The Game

there are a couple problems with our current system for joining ships. 1: memos are important, but people never use them. 2: some people want finer-grained control over their ship's RP level or crew, but they have no way of doing this. ship owners solve both problems, and the small misc changes i made all seem like common sense to me. 

## Changelog
:cl:
add: Added ship owners, an OOC utility to give players more control over their ships' RP levels. Any crew member can be the ship owner, as long as they are still connected and not AFK. This system is not ckey-based; if you respawn, you are no longer a candidate. Memos and job slot adjustment have been moved from the cryo console to the shipowner panel.
add: Ship applications -- ships can now enter an "application only" mode, where new players must send a ckey-linked application to the ship owner for approval or denial. These applications can be deleted by the owner afterwards; no other interaction is presently possible.
tweak: Ship job slot maximums and job slot change cooldown timers have been adjusted slightly.
refactor: "Latejoin" spawn code has been made more reliable and better able to handle a few edge cases.
admin: Admins can access a ship's ship owner panel from the shuttle manipulator, as well as directly access the docking port and ship overmap datum's VV panels.
/:cl:

### CODE REVIEW

**Comment by tmtmtl30** [comment]
shut the fuck up

File: code/controllers/subsystem/shuttle.dm [hunk 1]
```diff
@@ -461,6 +461,30 @@ SUBSYSTEM_DEF(shuttle)
 						. = TRUE
 						break
 
+		if("owner")
+			var/obj/docking_port/mobile/port = locate(params["id"]) in mobile
+			if(!port || !port.current_ship)
+				return
+			var/datum/overmap/ship/controlled/port_ship = port.current_ship
+			var/datum/action/ship_owner/admin/owner_action = new(port_ship)
+			owner_action.Grant(user)
```
Comment (ZephyrTFA) [inline code/controllers/subsystem/shuttle.dm:469]: Won't this end up giving admins another action every time, even if they already had it

Comment (tmtmtl30) [reply code/controllers/subsystem/shuttle.dm:469]: yeah, but the action automatically deletes itself when they close the window. so i don't think it's a big problem

Comment (ZephyrTFA) [reply code/controllers/subsystem/shuttle.dm:469]: Huh maybe I should actually read the entire PR before making review comments

**Comment by tmtmtl30** [comment]
switched things so that you now get prompted with whether or not you want to send your ckey with your application. i should probably make a more fleshed-out UI for this too but i don't know what the TGUI element for a big old text square is. that's pretty much the only thing stopping me

File: code/controllers/subsystem/server_maint.dm [hunk 2]
```diff
@@ -55,15 +55,16 @@ SUBSYSTEM_DEF(server_maint)
 	var/round_started = SSticker.HasRoundStarted()
 
 	var/kick_inactive = CONFIG_GET(flag/kick_inactive)
-	var/afk_period
-	if(kick_inactive)
-		afk_period = CONFIG_GET(number/afk_period)
-	for(var/I in currentrun)
-		var/client/C = I
+	var/afk_period = CONFIG_GET(number/afk_period)
+	for(var/client/C as anything in currentrun)
```
Comment (Kapu1178) [inline code/controllers/subsystem/server_maint.dm:59]: you cannot assume clients arent null

File: code/controllers/subsystem/shuttle.dm [hunk 3]
```diff
@@ -461,6 +461,30 @@ SUBSYSTEM_DEF(shuttle)
...
+			owner_action.Trigger()
+			return TRUE
+
+		if("vv_port")
+			var/obj/docking_port/mobile/port = locate(params["id"]) in mobile
+			if(!port)
+				return
+			user.client.debug_variables(port)
```
Comment (Kapu1178) [inline code/controllers/subsystem/shuttle.dm]: cant assume clients arent null

File: code/controllers/subsystem/shuttle.dm [hunk 4]
```diff
@@ -461,6 +461,30 @@ SUBSYSTEM_DEF(shuttle)
...
+			return TRUE
+
+		if("vv_ship")
+			var/obj/docking_port/mobile/port = locate(params["id"]) in mobile
+			if(!port || !port.current_ship)
+				return
+			user.client.debug_variables(port.current_ship)
```
Comment (Kapu1178) [inline code/controllers/subsystem/shuttle.dm]: cant assume clients arent null

**Review by Kapu1178 — CHANGES_REQUESTED** [review]

**Comment by tmtmtl30** [comment]
ok i figured out what the TGUI element for a big old text square is

**Comment by RDS88-beeman** [comment]
I'm really having problems with the new application feature. From my gameplay experience so far, if you want to join a ship with more than two or three people on it, you'll need to fill out an application. And since there's no way to give like a form to fill in with basic questions, you're forced to freehand it.

I cannot freehand an application for the life of me, especially when I don't know what the person wants. This mostly results in me sitting on the lobby screen, unwilling to join a ship with no crew but unable to join a ship with a lot of crew, wishing I could just join a ship and play shiptest like I used to.

It doesn't really help that from my experience there's no feedback if an application has even been viewed, let alone denied. I'd have to *hope* there is feedback for an accepted application, but I can't read code to I can't check.

As far as I can tell, all applications do to the game is section it up into the few people that can write apps all on one or two ships (Or it's just a metagang and the app is for the purpose of checking ckeys) and then people bored enough to actually join the game off into their own one and two person experiences, which is exactly the reason ships like the pill got removed.

TL:DR Pill syndrome has become an epidemic

**Comment by ZephyrTFA** [comment]
Turns out I accidentally merged this into master in a previous PR, which would be why everything is conflicting
